### PR TITLE
feat: Adds booking API data structure

### PR DESCRIPTION
**Pull request by JirkaChadima**
This PR aims to flesh out the booking data structure. It will get polished a bit (mandatory fields, descriptions etc.), the general idea will remain the same though. The latest version should be visible on https://app.swaggerhub.com/apis-docs/jirkachadima/wt-booking-api/will-be-set-at-runtime

**What is missing:**

- paid amenities or additional paid services
- billing information
- guarantee

@hynek-urban @czervenka Am I missing something important?

**TODO**

- [x] simplify the whole structure a bit by introducing guestInfo's id field
- [x] make either phone or email exclusive
- [ ] success response format

### CODE REVIEW

**Comment by JirkaChadima** [comment]
Thanks for the comment

1. I'll add it, the error list is not exhaustive ATM. I'll check with the code before submitting the final draft.
2. Do you mean `reseller.reference`? If you are an OTA, it might be useful to pair your record with the hotel record and vice versa.
3. I agree, but I don't know how to do `one of the two is required` in OAS.
4. Because I haven't done it yet :-) I'll add a revision that deals with that in an hour or so.
5. It's redundant. I think that in some cases `guestInfo` might actually not be necessary. You would then be unable to tell the actual number of guests.

**Comment by hynek-urban** [comment]
@JirkaChadima 

Ad 3, I suppose you can do it using a combination of "allOf" (for specifying the common base) and "oneOf" (to enumerate the variations of the individual variants).

Otherwise, I guess we're on the same page.

**Comment by JirkaChadima** [comment]
Thanks for the tip, I think I managed to make those two options exclusive. I think it's ready for merging unless @czervenka has any ideas on how to improve it.

**Comment by JirkaChadima** [comment]
TODO

- [ ] simplify the whole structure a bit by introducing guestInfo's `id` field
- [ ] make either phone or email exclusive

**Comment by JirkaChadima** [comment]
@hynek-urban @czervenka Ready for review again, if we're happy with the result I'll squash the commits and merge this.

**Comment by hynek-urban** [comment]
@JirkaChadima I have slightly changed the hierarchy of customer components, mostly because in the original version, customers with *both* phone and e-mail would be invalid (which is a property of the "oneOf" operator). Other than that, I think this is good to go!

File: docs/swagger.yaml [hunk 1]
```diff
@@ -30,3 +30,278 @@ paths:
...
+                description: Unique identifier for every guest, it can for example be a numerical ordinal. Every booking originator has to generate this data as it's used as a base for all other guest-related information (such as total number of guests).
+                type: string
+              name:
+                type: string
+              surname:
+                type: string
+              age:
+                description: Age at the time of arrival
+                type: number
+        roomTypes:
```
Comment (czervenka) [inline docs/swagger.yaml]: Shouldn't this be "rooms"? It specifies list of booked rooms not roomTypes?

File: docs/swagger.yaml [hunk 2]
```diff
@@ -30,3 +30,278 @@ paths:
...
+          type: array
+          description: List of rooms that form this booking. If a single room type is booked more than once, it shall be present more than once.
+          items:
+            $ref: '#/components/schemas/RoomType'
+    RatePlans:
+      description: List of rate plans that form the total price.
+      type: array
+      items:
+        $ref: '#/components/schemas/RatePlan'
+    RatePlan:
```
Comment (czervenka) [inline docs/swagger.yaml]: Why did you decided to send ratePlans with booking request?

File: docs/swagger.yaml [hunk 3]
```diff
@@ -30,3 +30,278 @@ paths:
...
+          type: string
+          format: date
+        guestInfoIds:
+          type: array
+          items:
+            description: Which guests will be staying in this room? The strings should be id fields from guestInfo, so we can match the guests and their additional information when needed.
+            type: string
+        appliedModifier:
+          $ref: 'https://raw.githubusercontent.com/windingtree/wiki/ac6c33d38a9f91aa4b951a00acacf11905208550/hotel-data-swagger.yaml#/components/schemas/RatePlanPriceModifier'
+        subtotal:
```
Comment (czervenka) [inline docs/swagger.yaml]: We should make it clear whether this is total price for all gusts based on the rate plan or a price listed in the rate plan by one guest.

File: docs/swagger.yaml [hunk 4]
```diff
@@ -30,3 +30,278 @@ paths:
...
+      summary: Create booking
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/Booking'
+      responses:
+        200:
+          description: Success
```
Comment (czervenka) [inline docs/swagger.yaml]: Theres should be success response body format, shouldn't it?

File: docs/swagger.yaml [hunk 5]
```diff
@@ -30,3 +30,278 @@ paths:
...
+      description: Booking object passed from a customer to a hotel
+      type: object
+      required:
+        - customer
+        - reseller
+        - hotelId
+        - booking
+        - pricing
+      properties:
+        createdAt:
```
Comment (czervenka) [inline docs/swagger.yaml]: What is createdAt necessary for? Let's remove it.

File: docs/swagger.yaml [hunk 6]
```diff
@@ -30,3 +30,278 @@ paths:
...
+    BookingInfo:
+      description: Information on the booked property
+      type: object
+      required:
+        - arrival
+        - departure
+        - roomTypes
+        - guestInfo
+      properties:
+        arrival:
```
Comment (czervenka) [inline docs/swagger.yaml]: This format expects that all guests will arrive at the same day and stay the same number of days. It is common that some of the guests has shorter stay or arrives sooner or later, AFAIK.

File: docs/swagger.yaml [hunk 7]
```diff
@@ -30,3 +30,325 @@ paths:
...
+            schema:
+              $ref: '#/components/schemas/Booking'
+      responses:
+        200:
+          description: Success
+        400:
+          $ref: '#/components/responses/BadRequestError'
+        409:
+          $ref: '#/components/responses/ConflictError'
+        422:
```
[the remark this answers is not in the document]
Comment (JirkaChadima) [reply docs/swagger.yaml]: This is compatible with all of the other APIs we've done so far. I'd keep it.

File: docs/swagger.yaml (hunk 4 above)
Comment (czervenka) [reply docs/swagger.yaml]: I don't know about anything else now.

File: docs/swagger.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (czervenka) [reply docs/swagger.yaml]: Makes sense.

File: docs/swagger.yaml (hunk 6 above)
Comment (czervenka) [reply docs/swagger.yaml]: You are right. We don't need to fix it now and we need to have a prototypical price calculation example for others first. Thanks.

**Pull request merged**